gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/user/password/PasswordHistoryHandler.java

```diff
@@ -0,0 +1,71 @@
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.am.gateway.handler.root.resources.handler.user.password;
+
+import io.gravitee.am.common.jwt.JWT;
+import io.gravitee.am.common.utils.ConstantKeys;
+import io.gravitee.am.gateway.handler.root.service.user.UserService;
+import io.gravitee.am.model.Domain;
+import io.gravitee.am.model.ReferenceType;
+import io.gravitee.am.model.User;
+import io.gravitee.am.model.oidc.Client;
+import io.gravitee.am.service.exception.PasswordHistoryException;
+import io.gravitee.am.service.impl.PasswordHistoryService;
+import io.gravitee.common.http.HttpStatusCode;
+import io.vertx.core.Handler;
+import io.vertx.reactivex.ext.web.RoutingContext;
+
+/**
+ * Checks a password against a user's history of passwords and returns either a
+ * 200 ok or 401 failure if the password is already in the history.
+ */
+public class PasswordHistoryHandler implements Handler<RoutingContext> {
+
+    private final PasswordHistoryService passwordHistoryService;
+    private final UserService userService;
+    private final Domain domain;
+
+    public PasswordHistoryHandler(PasswordHistoryService passwordHistoryService, UserService userService, Domain domain) {
+        this.passwordHistoryService = passwordHistoryService;
+        this.userService = userService;
+        this.domain = domain;
+    }
+
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    @Override
+    public void handle(RoutingContext context) {
+        User user1 = context.get(ConstantKeys.USER_CONTEXT_KEY);
+//        Client client = context.get(ConstantKeys.CLIENT_CONTEXT_KEY);
+        var accessToken = context.request().getFormAttribute(ConstantKeys.TOKEN_CONTEXT_KEY);
+        var password = context.request().getFormAttribute(ConstantKeys.PASSWORD_PARAM_KEY);
+        userService.verifyToken(accessToken)
+                   .doOnError(throwable -> context.fail(HttpStatusCode.INTERNAL_SERVER_ERROR_500, throwable))
+                   .subscribe(userToken -> {
+                       var user = userToken.getUser();
+                       passwordHistoryService
+                               .passwordAlreadyUsed(ReferenceType.DOMAIN, domain.getId(), user.getId(), password, domain.getPasswordSettings())
+                               .subscribe(passwordUsed -> {
+                                   if (Boolean.TRUE.equals(passwordUsed)) {
+                                       context.fail(HttpStatusCode.BAD_REQUEST_400, PasswordHistoryException.passwordAlreadyInHistory(domain.getPasswordSettings()));
```

Comment:
We should use the PasswordSettings.getSettings method to gety the right password policy rules (app or domain)